#  CONTENTS LIST
## 1. Importing libraries
## 2. Import datafrmaes
## 3. look for Mixed-Type Data
## 4. look for missing values
## 5. look for duplicates
## 6. Tidying Up and Exporting Changes
## 7. Looking for annomalies
## 8. check for mixed types
## 9. Finding missing values
## 10. check for dulicates
## 11. Export dataframes

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#create path
path = r'C:\Users\Christian\Desktop\Mein Ordner\CF\Data Immersion\Achievement4\Instacart Basket Analysis'

In [4]:
# import dataframes

In [5]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders.csv'), index_col = False)

In [6]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [7]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Mixed-Type Data

In [8]:
df_test = pd.DataFrame()

In [9]:
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#check for mixed data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
#check for mixed data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

# Missing Values

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [20]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

# Tidying Up and Exporting Changes

In [25]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

## 2.

In [26]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

The only thing that seem a little bit strange to me that the minimum in tha days_since_prior_order column is 0 which would mean
that someone made more than one order in one day. Beside that I don't see anything strange that would need further investigation.

In [27]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [47]:
df_prods[df_prods['prices'] >  30]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

The anmomaly is clearly the price max value of 99999. This is very unlikely since the 3rd quartile is 11.2.
The most obvious explanation is that the value was entered incorrectly. This explanation is strenghtened by the fact that 
there are only two values bigger than 30 and they are very big outliers.


## 3.

In [28]:
#check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
  else: print(col + ': ' ' NO MIXED VALUES FOUND !')

order_id:  NO MIXED VALUES FOUND !
user_id:  NO MIXED VALUES FOUND !
eval_set:  NO MIXED VALUES FOUND !
order_number:  NO MIXED VALUES FOUND !
order_dow:  NO MIXED VALUES FOUND !
order_hour_of_day:  NO MIXED VALUES FOUND !
days_since_prior_order:  NO MIXED VALUES FOUND !


## 4.

No mix data types were found.

## 5.

In [29]:
# Finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In the column days_since_prior_order there are 206209 missing values. The only explanation I can think of is that these are the 
customers who made their first orders so that there is no prior day.

## 6.

In [30]:
# check for missing vaules
df_null = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [31]:
df_null

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

With a substantial volume of 206209 observations containing missing values, it is crucial to approach the issue cautiously. 
Simply deleting or imputing these values could introduce inaccuracies into the subsequent analysis. 
The best decision is to retain the missing data in its original state. 
This choice acknowledges the unique context of the first order for each user and ensures that the analysis is 
conducted with consideration for the inherent nature of these initial transactions.

## 7.

In [32]:
# check for dulicates
df_ords_dup = df_ords[df_ords.duplicated()]

In [33]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates were found.

## 8.

Since there are no duplicates no further action is needed.

## 9.

In [34]:
# Export to csv

In [35]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

In [36]:
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_checked.csv'))